src/Data/Nat/Properties.agda

```diff
@@ -42,15 +44,33 @@ open ≡-Reasoning
 suc-injective : ∀ {m n} → suc m ≡ suc n → m ≡ n
 suc-injective refl = refl
 
-infix 4 _≟_
+≡ᵇ⇒≡ : ∀ m n → T (m ≡ᵇ n) → m ≡ n
+≡ᵇ⇒≡ zero    zero    _  = refl
+≡ᵇ⇒≡ (suc m) (suc n) eq = cong suc (≡ᵇ⇒≡ m n eq)
+≡ᵇ⇒≡ zero    (suc n) ()
+≡ᵇ⇒≡ (suc m) zero    ()
+
+≡⇒≡ᵇ : ∀ m n → m ≡ n → T (m ≡ᵇ n)
+≡⇒≡ᵇ zero    zero    eq = _
+≡⇒≡ᵇ (suc m) (suc n) eq = ≡⇒≡ᵇ m n (suc-injective eq)
+≡⇒≡ᵇ zero    (suc n) ()
+≡⇒≡ᵇ (suc m) zero    ()
+
+-- NB: we use the builtin function `_≡ᵇ_ : (m n : ℕ) → Bool` here so
+-- that the function quickly decides whether to return `yes` or `no`.
+-- It sill takes a linear amount of time to generate the proof if it
+-- is inspected. We expect the main benefit to be visible in compiled
+-- code: the backend erases proofs.
 
+infix 4 _≟_
 _≟_ : Decidable {A = ℕ} _≡_
-zero  ≟ zero   = yes refl
-zero  ≟ suc n  = no λ()
-suc m ≟ zero   = no λ()
-suc m ≟ suc n  with m ≟ n
-... | yes refl = yes refl
-... | no m≢n   = no (m≢n ∘ suc-injective)
+m ≟ n = map′ (≡ᵇ⇒≡ m n) (≡⇒≡ᵇ m n) (T? (m ≡ᵇ n))
+
+≡-irrelevant : Irrelevant {A = ℕ} _≡_
+≡-irrelevant = Decidable⇒UIP.≡-irrelevant _≟_
+
+≟-diag : ∀ {m n} (eq : m ≡ n) → (m ≟ n) ≡ yes eq
+≟-diag = ≡-≟-identity _≟_
```

Comment:
Just noticed this proof. I'm not sure this one should be here? There's nothing particular to natural numbers in it.

Reply:
I'd argue it's good to have a specialised version available.